hazelcast-hibernate/hazelcast-hibernate3/src/main/java/com/hazelcast/hibernate/HazelcastTimestamper.java

```diff
@@ -29,13 +32,42 @@
 public final class HazelcastTimestamper {
 
     private static final int SEC_TO_MS = 1000;
+    /**
+     * Value for left shifting instance.getCluster().getClusterTime(), freeing some space for the counter.
+     * Right now value is constant set to 12.
+     */
+    private static final int BASE_TIME_SHIFT = 12;
+    private static final int ONE_MS = 1 << BASE_TIME_SHIFT;
+    private static final AtomicLong VALUE  = new AtomicLong();
+    private final ILogger logger = Logger.getLogger(getClass());
 
     private HazelcastTimestamper() {
     }
 
+    /**
+     * Returns an increasing unique value based on the cluster time. First loop works like a thread spin-waits and
+     * inner for loop guarantees that unique value is generated.
+     * @see com.hazelcast.hibernate.HazelcastTimestamper#BASE_TIME_SHIFT
+     * @return uniquely & increasing value
+     */
     public static long nextTimestamp(HazelcastInstance instance) {
-        // System time in ms.
-        return instance.getCluster().getClusterTime();
+        int runs = 0;
+        while (true) {
+            long base = instance.getCluster().getClusterTime() << BASE_TIME_SHIFT;
```

Comment:
Yes you are right, it may cause an infinite loop scenario, however inside the for loop, if statement will return the value no matter what,
```
 if (VALUE.compareAndSet(current, update)) {
         if (runs > 1) {
               Logger.getLogger(HazelcastTimestamper.class)
                    .finest(String.format("Waiting for time to pass. Looped %d times", runs));
          }
           return update;
  }
```
because update = Math.max(base, current + 1) will assign update to current + 1 and method will return inside if statement because VALUE.compareAndSet(current, update) will return true because update will be current + 1. Again, you are right the loop can be constructed as an infinite loop because update < maxValue will never be satisfied, however the method will always return in the first run of the loop.